### PR TITLE
[AC-1218] Add ability to delete Provider Portals

### PR DESCRIPTION
**Pull request by cyprain-okeke**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Currently, there is no ability to delete Provider Portals. This functionality should be available from the internal Bitwarden Portal.

On the Provider page, there will be an option to delete the Provider at the bottom of the page. A Provider can only be deleted if no organizations are linked to the Provider.

Deleting from

[Bitwarden portal](https://www.figma.com/proto/m8eMk6yeKfOq4AUuyp70ih/Consolidated-Billing?page-id=698%3A10060&type=design&node-id=1171-18005&viewport=-11087%2C521%2C0.25&t=tpHXkqSav41X2FCa-1&scaling=min-zoom&starting-point-node-id=1399%3A12380&show-proto-sidebar=1&mode=design)

[Provider portal - attempting to delete with clients](https://www.figma.com/proto/m8eMk6yeKfOq4AUuyp70ih/Consolidated-Billing?page-id=698%3A10060&type=design&node-id=1309-18496&viewport=-11087%2C521%2C0.25&t=tpHXkqSav41X2FCa-1&scaling=min-zoom&starting-point-node-id=1309%3A18496&show-proto-sidebar=1&mode=design)

[Provider portal - deleting without clients](https://www.figma.com/proto/m8eMk6yeKfOq4AUuyp70ih/Consolidated-Billing?page-id=698%3A10060&type=design&node-id=1710-15961&viewport=-11087%2C521%2C0.25&t=tpHXkqSav41X2FCa-1&scaling=min-zoom&starting-point-node-id=1710%3A15961&show-proto-sidebar=1&mode=design)

## Code changes

* **file.ext:** Description of what was changed and why

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
Screenshot 2024-04-15 at 11 06 42
Screenshot 2024-04-15 at 11 07 06
Screenshot 2024-04-15 at 11 07 14
Screenshot 2024-04-15 at 11 08 26
Screenshot 2024-04-15 at 11 09 31
Screenshot 2024-04-15 at 11 09 38
Screenshot 2024-04-15 at 11 10 59

### CODE REVIEW

File: src/Admin/AdminConsole/Controllers/ProvidersController.cs [hunk 1]
```diff
@@ -275,4 +277,64 @@ public async Task<IActionResult> CreateOrganization(Guid providerId, Organizatio
...
+
+            if (providerName.ToLower().Trim() != provider.Name.ToLower().Trim())
+            {
+                return BadRequest("Invalid provider name");
+            }
+
+            await _providerService.DeleteAsync(provider);
+        }
+
+        return NoContent();
```
Comment (amorask-bitwarden) [inline src/Admin/AdminConsole/Controllers/ProvidersController.cs:305]: ❓ In my opinion, responding with No Content here gives the impression that this operation was successful when in reality, it was invalid. How is this handled by the front-end?

File: src/Admin/AdminConsole/Controllers/ProvidersController.cs [hunk 2]
```diff
@@ -275,4 +277,64 @@ public async Task<IActionResult> CreateOrganization(Guid providerId, Organizatio
...
+    }
+
+    [HttpPost]
+    [SelfHosted(NotSelfHostedOnly = true)]
+    [RequirePermission(Permission.Provider_Edit)]
+    public async Task<IActionResult> DeleteInitiation(Guid id, string providerEmail)
```
Comment (amorask-bitwarden) [inline src/Admin/AdminConsole/Controllers/ProvidersController.cs:311]: ❓ Do we need to do the same check to see if the provider has any `ProviderOrganization` records linked to it as we do in the `Delete` endpoint?

File: src/Admin/AdminConsole/Controllers/ProvidersController.cs [hunk 3]
```diff
@@ -275,4 +277,64 @@ public async Task<IActionResult> CreateOrganization(Guid providerId, Organizatio
 
         return RedirectToAction("Edit", "Providers", new { id = providerId });
     }
+
+    [HttpPost]
+    [SelfHosted(NotSelfHostedOnly = true)]
+    [RequirePermission(Permission.Provider_Edit)]
+    public async Task<IActionResult> Delete(Guid id, string providerName)
```
Comment (amorask-bitwarden) [inline src/Admin/AdminConsole/Controllers/ProvidersController.cs:284]: When I try to delete a `Provider` in a `Pending` state from Admin, I get the following error:
image

**Review by amorask-bitwarden — CHANGES_REQUESTED** [review]

File: src/Admin/AdminConsole/Controllers/ProvidersController.cs (hunk 3 above)
Comment (cyprain-okeke) [reply src/Admin/AdminConsole/Controllers/ProvidersController.cs:284]: this has been resolved

File: src/Admin/AdminConsole/Controllers/ProvidersController.cs (hunk 2 above)
Comment (cyprain-okeke) [reply src/Admin/AdminConsole/Controllers/ProvidersController.cs:308]: The front end is validating if the provider has a client and if only when there is no client that his method is called.

File: src/Admin/AdminConsole/Controllers/ProvidersController.cs (hunk 1 above)
Comment (cyprain-okeke) [reply src/Admin/AdminConsole/Controllers/ProvidersController.cs:302]: This has been refactored. but The front end is validating if the provider has a client and if only when there is no client that his method is called.

**Review by amorask-bitwarden — APPROVED** [review]

File: src/Core/Auth/Models/Business/Tokenables/ProviderDeleteTokenable.cs
Comment (jlf0dev) [inline src/Core/Auth/Models/Business/Tokenables/ProviderDeleteTokenable.cs:1]: Does this belong in Auth? Should we move it to AC?

Comment (cyprain-okeke) [reply src/Core/Auth/Models/Business/Tokenables/ProviderDeleteTokenable.cs:1]: moved to adminConsole

Comment (jlf0dev) [reply src/Core/Auth/Models/Business/Tokenables/ProviderDeleteTokenable.cs:1]: Thanks, should just need AC approval now.

File: src/Admin/AdminConsole/Controllers/ProvidersController.cs [hunk 4]
```diff
@@ -275,4 +277,76 @@ public async Task<IActionResult> CreateOrganization(Guid providerId, Organizatio
...
+    }
+
+    private static bool IsValidEmail(string email)
+    {
+        const string pattern = @"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$";
+        var regex = new Regex(pattern);
+        return regex.IsMatch(email);
+    }
+
+    private string ValidateProviderName(Provider provider, string providerName)
```
Comment (r-tome) [inline src/Admin/AdminConsole/Controllers/ProvidersController.cs:338]: This method is only used once. I suggest moving this logic into the controller as it also checks if the Provider is null

File: src/Api/AdminConsole/Controllers/ProvidersController.cs [hunk 5]
```diff
@@ -123,4 +123,41 @@ public async Task<ProviderResponseModel> Setup(Guid id, [FromBody] ProviderSetup
...
+        if (provider == null)
+        {
+            throw new NotFoundException();
+        }
+        await _providerService.DeleteAsync(provider, model.Token);
+    }
+
+    [HttpDelete("{id}")]
+    [HttpPost("{id}/delete")]
+    public async Task Delete(string id)
```
Comment (r-tome) [inline src/Api/AdminConsole/Controllers/ProvidersController.cs:141]:
```suggestion
    public async Task Delete(Guid id)
```

**Review by r-tome — CHANGES_REQUESTED** [review]
Looking good, there's some changes I'd like to see

File: bitwarden_license/src/Commercial.Core/AdminConsole/Services/ProviderService.cs [hunk 6]
```diff
@@ -594,6 +602,44 @@ public async Task LogProviderAccessToOrganizationAsync(Guid organizationId)
...
+    {
+        var providerAdmin = await _userRepository.GetByEmailAsync(providerAdminEmail);
+        if (providerAdmin == null)
+        {
+            throw new BadRequestException("Provider admin not found.");
+        }
+        if (string.IsNullOrWhiteSpace(provider.Name))
+        {
+            throw new BadRequestException("Provider name not found.");
+        }
```
[the remark this answers is not in the document]
Comment (cyprain-okeke) [reply bitwarden_license/src/Commercial.Core/AdminConsole/Services/ProviderService.cs:615]: resolved

File: src/Admin/AdminConsole/Controllers/ProvidersController.cs [hunk 7]
```diff
@@ -275,4 +277,76 @@ public async Task<IActionResult> CreateOrganization(Guid providerId, Organizatio
...
+        await _providerService.DeleteAsync(provider);
+        return NoContent();
+    }
+
+    [HttpPost]
+    [SelfHosted(NotSelfHostedOnly = true)]
+    [RequirePermission(Permission.Provider_Edit)]
+    public async Task<IActionResult> DeleteInitiation(Guid id, string providerEmail)
+    {
+        if (!IsValidEmail(providerEmail))
```
[the remark this answers is not in the document]
Comment (cyprain-okeke) [reply src/Admin/AdminConsole/Controllers/ProvidersController.cs:310]: resolve

File: src/Admin/AdminConsole/Controllers/ProvidersController.cs (hunk 4 above)
Comment (cyprain-okeke) [reply src/Admin/AdminConsole/Controllers/ProvidersController.cs:338]: resolved

File: src/Api/AdminConsole/Controllers/ProvidersController.cs (hunk 5 above)
Comment (cyprain-okeke) [reply src/Api/AdminConsole/Controllers/ProvidersController.cs:141]: resolved

File: bitwarden_license/test/Commercial.Core.Test/AdminConsole/Services/ProviderServiceTests.cs
[the remark this answers is not in the document]
Comment (cyprain-okeke) [reply bitwarden_license/test/Commercial.Core.Test/AdminConsole/Services/ProviderServiceTests.cs:1]: Add some unit test

**Review by r-tome — APPROVED** [review]
Good work!

**Pull request merged**